### PR TITLE
Add log rolling support for the file logger

### PR DESCRIPTION
**Pull request by MohamedM216**
# Issue

This PR adds log rolling support for the `FileLogger` class. It addresses this TODO:https://github.com/valhalla/valhalla/blob/49985ad593df413f62dd54555a389670a67f9dcf/src/midgard/logging.cc#L194

Note: we can't reconfigure the current logger, so to make tests pass we must run each test separately. So far I have 2 ideas in mind to be able to run all tests together: 
- make the current logger configurable
- or we can combine all tests in one test

Let me know your thoughts.

## Tasklist

 - [x] Add tests
 - [ ] Add #fixes with the issue number that this PR addresses
 - [ ] Update the docs with any new request parameters or changes to behavior described
 - [x] Update the [changelog](CHANGELOG.md)
 - [ ] If you made changes to the lua files, update the [taginfo](taginfo.json) too
 - [ ] If you made changes to a translation file, [update transifex](docs/docs/locales.md) too

## Requirements / Relations

 Link any requirements here. Other pull requests this PR is based on?

### CODE REVIEW

File: src/midgard/logging.cc [hunk 1]
```diff
@@ -215,13 +216,48 @@ class FileLogger : public Logger {
       }
     }
 
+    // Configure log rolling
+    max_file_size = 50 * 1024 * 1024; // 50 MB
+    auto size_config = config.find("max_file_size");
```
Comment (nilsnolde) [inline src/midgard/logging.cc:219]: you'll also need to add this to `valhalla_build_config`

File: src/midgard/logging.cc [hunk 2]
```diff
@@ -215,13 +216,48 @@ class FileLogger : public Logger {
...
+    if (size_config != config.end()) {
+      try {
+        if (!size_config->second.empty() && size_config->second[0] == '-') {
```
Comment (nilsnolde) [inline src/midgard/logging.cc]: IMO an empty value `""` should just be signalling that one doesn't want this feature

File: src/midgard/logging.cc [hunk 3]
```diff
@@ -215,13 +216,48 @@ class FileLogger : public Logger {
...
+          throw std::runtime_error("max_file_size must be greater than 0, got: " +
+                                   size_config->second);
+        }
+        max_file_size = std::stoull(size_config->second);
+        if (max_file_size == 0) {
+          throw std::runtime_error("max_file_size must be greater than 0");
+        }
+      } catch (...) { throw std::runtime_error(size_config->second + " is not a valid file size"); }
+    }
```
Comment (nilsnolde) [inline src/midgard/logging.cc]: to make this a bit less code you could just parse it into an int64 and check if it's > 0

File: src/midgard/logging.cc [hunk 4]
```diff
@@ -236,6 +272,53 @@ class FileLogger : public Logger {
   }
 
 protected:
+  void CheckRollLog() {
+    lock.lock();
+    if (file.is_open()) {
+      std::streampos current_size = file.tellp();
+      if (current_size > static_cast<std::streampos>(max_file_size)) {
+        lock.unlock();
```
Comment (nilsnolde) [inline src/midgard/logging.cc]: why unlock here only to lock again first thing in RollLog? could do a lock_guard in this function and remove the other one?

File: src/midgard/logging.cc [hunk 5]
```diff
@@ -269,10 +352,22 @@ class FileLogger : public Logger {
...
+      }
+    }
+  }
+
   std::string file_name;
   std::ofstream file;
   std::chrono::seconds reopen_interval;
   std::chrono::system_clock::time_point last_reopen;
+  std::size_t max_file_size;
+  unsigned int max_archived_files;
```
Comment (nilsnolde) [inline src/midgard/logging.cc]: could do the init to 50 & 10 here?

File: src/midgard/logging.cc [hunk 6]
```diff
@@ -236,6 +272,53 @@ class FileLogger : public Logger {
...
+
+  void RollLog() {
+    std::lock_guard<std::mutex> lock_guard(lock);
+
+    if (file.is_open()) {
+      file.close();
+    }
+
+    // Roll existing files (archive.2 -> archive.3, archive.1 -> archive.2, etc.)
+    for (int i = max_archived_files - 1; i > 0; --i) {
```
Comment (nilsnolde) [inline src/midgard/logging.cc]: small optimization: can we re-write this to start at 1 instead of `max_archived_files`? then we can break as soon as we can't find any more `old_file` so it won't do 999 no-op loops for 1000 max files during the first rolling.

File: src/midgard/logging.cc [hunk 7]
```diff
@@ -269,10 +352,22 @@ class FileLogger : public Logger {
...
   }
+
+  void EnsureDirectoryExists() {
+    const auto parent_dir = std::filesystem::path(file_name).parent_path();
+    if (!parent_dir.empty() && !std::filesystem::is_directory(parent_dir)) {
+      if (!std::filesystem::create_directories(parent_dir)) {
+        throw std::runtime_error("Cannot create directory for log file: " + parent_dir.string());
+      }
+    }
+  }
```
Comment (nilsnolde) [inline src/midgard/logging.cc:360]: this exact code is copied from a few dozen lines further up, better remove that and call this function

**Comment by nilsnolde** [comment]
also we should add a note in `valhalla_build_config`'s help strings that the file logger is unsuitable for multi-processing mode and this log rolling especially, there's a race condition which will very likely happen and might crash the service.

**Comment by nilsnolde** [comment]
> Note: we can't reconfigure the current logger, so to make tests pass we must run each test separately.

> We can't combine all tests to be just one test because we need to reconfigure the logger more than once. So, I suggest updating the logger to be reconfigurable -to be able to run all tests together- unless its nature must not be changed.

I think making it reconfigurable is a performance issue, we need logging to be super fast without lots of runtime branches. I'd suggest you add some more testing code to the currently existing test, i.e. also change the existing config to add the rolling stuff (only the happy path). We're trying not to add too many more test files, it just adds too much compilation time.

File: src/midgard/logging.cc (hunk 2 above)
Comment (MohamedM216) [reply src/midgard/logging.cc]: OK! After some updates, here is the expected behavior:

- If both present and valid => enable log rolling.
- If only one present and valid and the other present and empty => don’t throw, just don’t enable.
- If one is invalid (empty, <= 0 or non-numeric) → throw.

File: test/logging_rolling.cc [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+#include <chrono>
+#include <cstdio>
+#include <filesystem>
+#include <string>
+#include <thread>
+
+using namespace valhalla::midgard;
+
+namespace {
+TEST(Logging, FileLoggerRollingTest) {
```
Comment (MohamedM216) [inline test/logging_rolling.cc]: Combining logic into one test block didn't work as we must reconfigure the logger to test both scenarios (`"reopen_interval"` and log rolling). So, I created this new test file for this path (just the happy path).

File: src/midgard/logging.cc [hunk 9]
```diff
@@ -269,10 +347,23 @@ class FileLogger : public Logger {
...
+
   std::string file_name;
   std::ofstream file;
   std::chrono::seconds reopen_interval;
   std::chrono::system_clock::time_point last_reopen;
+  bool log_rolling_enabled = false;
+  std::size_t max_file_size = 50 * 1024 * 1024; // 50 MB default
+  unsigned int max_archived_files = 10;         // 10 files default
 };
```
Comment (nilsnolde) [inline src/midgard/logging.cc]: hm sorry, I should've said "initialize those values to 0". again, we don't want to enable this by default, which is what it's currently doing. in the config builder those values are optional but when they're not set (which is basically everyone who installs or even updates valhalla without explicitly setting those configs to 0), it'll enable log rolling.

File: src/midgard/logging.cc [hunk 10]
```diff
@@ -215,13 +216,43 @@ class FileLogger : public Logger {
...
+        throw std::runtime_error("max_archived_files must be greater than 0");
+      }
+      max_archived_files = parsed_archive;
+    }
+
+    // Enable log rolling only if both values are provided and valid
+    if ((size_config != config.end() && !size_config->second.empty()) &&
+        (archive_config != config.end() && !archive_config->second.empty())) {
+      log_rolling_enabled = true;
+    }
```
Comment (nilsnolde) [inline src/midgard/logging.cc]: once `max_file_size` & `max_archived_files` are init'd to 0, this can just become 

```suggestion
    log_rolling_enabled = (max_file_size > 0) && (max_archived_files > 0);
```

File: test/logging_rolling.cc (hunk 8 above)
Comment (nilsnolde) [inline test/logging_rolling.cc]: I wasn't very clear here I think, sorry. I meant: the current logging.cc test could be simply made a rolling logger and test the happy path there instead of a new file (more compilation etc). I think that'll be possible without compromising the current test it does?

File: scripts/valhalla_build_config [hunk 11]
```diff
@@ -408,19 +436,31 @@ help_text = {
...
         'logging': {
+            '__note__': (
+                'The file logger is not suitable for multi-processing mode. '
+                'Log rolling in particular may cause race conditions and crashes.'
+            ),
             'type': 'Type of logger either std_out or file',
             'color': 'User colored log level in std_out logger',
             'file_name': 'Output log file for the file logger',
             'long_request': 'Value used in processing to determine whether it took too long',
+            'max_file_size': 'Maximum file size in bytes before rolling to a new file ("" = disabled)',
```
Comment (nilsnolde) [inline scripts/valhalla_build_config]:
```suggestion
            'max_file_size': 'Maximum file size in bytes before rolling to a new file (0 = disabled)',
```

File: src/midgard/logging.cc [hunk 12]
```diff
@@ -215,13 +216,43 @@ class FileLogger : public Logger {
       }
     }
 
+    // Configure log rolling - default to disabled
+    auto size_config = config.find("max_file_size");
+    auto archive_config = config.find("max_archived_files");
+
+    if (size_config != config.end() && !size_config->second.empty()) {
+      auto parsed_size = std::stoll(size_config->second);
+      if (parsed_size <= 0) {
```
Comment (nilsnolde) [inline src/midgard/logging.cc]: why's 0 throwing? shouldn't that signal that this feature should be turned off?

File: test/logging_rolling.cc (hunk 8 above)
Comment (MohamedM216) [reply test/logging_rolling.cc]: If you mean adding a new test case that only covers the happy path of the log rolling feature alongside the existing one in `test/logging.cc`, that isn’t possible with the current logger implementation (because it uses the singleton pattern). This limitation is the same reason I had to introduce the log rolling test in a separate file.

However, if you mean replacing the existing test case with the log rolling one, then that is possible.

File: src/midgard/logging.cc (hunk 12 above)
Comment (MohamedM216) [reply src/midgard/logging.cc]: I just wanted to distinguish between 2 cases:
- the feature is disabled by default (which means the first 2 `if`s are false and both `max_file_size` and `max_archived_files` are 0) 
- and when the user includes `max_file_size` or `max_archived_files` in the configuration map but leaves one or both improperly set.

Comment (nilsnolde) [reply src/midgard/logging.cc]: this is not addressed yet..

File: test/logging_rolling.cc (hunk 8 above)
Comment (nilsnolde) [reply test/logging_rolling.cc]: this is also not addressed yet

**Review by nilsnolde — CHANGES_REQUESTED** [review]

**Comment by nilsnolde** [comment]
you requested a review @MohamedM216 but nothing changed. feel free to argue on the things I proposed.

**Comment by MohamedM216** [comment]
Oh! I think I figured out what happened. You couldn't see my last comments because they were pending. I forgot to submit the review. I think now you can see them:

https://github.com/valhalla/valhalla/pull/5477#discussion_r2350740834

https://github.com/valhalla/valhalla/pull/5477#discussion_r2310311640

https://github.com/valhalla/valhalla/pull/5477#discussion_r2348900995

Sorry about the confusion!

File: test/logging_rolling.cc (hunk 8 above)
Comment (nilsnolde) [reply test/logging_rolling.cc]: What's the problem with using `reopen_interval` **and** log rolling in a single configuration? I don't immediately see that they clash somehow (and if they did the code must protect against that)

File: src/midgard/logging.cc (hunk 12 above)
Comment (nilsnolde) [reply src/midgard/logging.cc]: > the feature is disabled by default (which means the first 2 ifs are false and both max_file_size and max_archived_files are 0)

which "first 2 ifs"? the way I read this: if they are set in the config with 0, then we'd throw. that's not desirable, since it's the default in the code and both things should end up doing the same: disable the feature.

File: src/midgard/logging.cc [hunk 13]
```diff
@@ -268,10 +343,23 @@ class FileLogger : public Logger {
...
+    }
+  }
+
   std::string file_name;
   std::ofstream file;
   std::chrono::seconds reopen_interval;
   std::chrono::system_clock::time_point last_reopen;
+  bool log_rolling_enabled = false;
+  std::size_t max_file_size = 0;
+  unsigned int max_archived_files = 0;
```
Comment (nilsnolde) [inline src/midgard/logging.cc]: these should be `int64_t` right (or at least smth _signed_)? we're (correctly) protecting against negative values and throw if that's in the config. but currently negative values would end up with huge positive values when `std::stoll` is used on them to parse them to unsigned ones and hence wouldn't throw, but end up allowing a zillion files with 1000s of GB each

Comment (MohamedM216) [reply src/midgard/logging.cc]: I chose to use `std::stoll` into `long long` (`parsed_size` / `parsed_archive`) first, so that I can explicitly check for negative values before assigning to `max_file_size` and `max_archived_files`. I also added tests for this case (passing negative values in the config), and the checks correctly throw exceptions, so we’re covered against the “huge positive number” issue you mentioned.

Comment (MohamedM216) [reply src/midgard/logging.cc]: I think we should keep them as they are. It’s very unlikely anyone would need billions of archived files or log files sized in billions of GB. But we still have another problem: narrowing conversions (`long long` -> `unsigned long` or `long long` -> `unsigned int`). If a user enters a value that fits within `LONG_LONG_MAX` but exceeds `INT_MAX` or `LONG_MAX`, it would overflow. 

To address this, I changed:
- `std::stoll` -> `std::stol` for `max_file_size`
- `std::stoll` -> `std::stoi` for `max_archived_files`

Addressed here: https://github.com/valhalla/valhalla/pull/5477/commits/946fa0dd5bfd41010366d828cfb3427383441224

File: src/midgard/logging.cc [hunk 14]
```diff
@@ -214,13 +215,48 @@ class FileLogger : public Logger {
...
+    auto size_config = config.find("max_file_size");
+    auto archive_config = config.find("max_archived_files");
+
+    if (size_config != config.end() && !size_config->second.empty()) {
+      try {
+        auto parsed_size = std::stol(size_config->second);
+        if (parsed_size < 0) {
+          throw std::runtime_error("max_file_size must be greater than 0");
+        }
+        max_file_size = parsed_size;
```
Comment (nilsnolde) [inline src/midgard/logging.cc]: sorry, you're right, you didn't parse them directly into the members. somehow I didn't see that.

however, we don't have to do this detour. you can parse directly into the member variables if you make them both signed integers. it's nitpicky, but every cycle counts in the logger:)

Comment (MohamedM216) [reply src/midgard/logging.cc]: addressed here: https://github.com/valhalla/valhalla/pull/5477/commits/537ea123b7405321de86172ee184eabe894109cf

File: src/midgard/logging.cc [hunk 15]
```diff
@@ -235,6 +269,58 @@ class FileLogger : public Logger {
...
+  }
+
+  void RollLog() {
+    if (file.is_open()) {
+      file.close();
+    }
+
+    // Find the highest existing archive file
+    int highest = 0;
+    for (int i = 1; i <= static_cast<int>(max_archived_files) - 1; ++i) {
```
Comment (MohamedM216) [inline src/midgard/logging.cc:290]: This was keeping `max_archived_files` + 1 max archived files. Now it is fixed.

File: test/logging.cc [hunk 16]
```diff
@@ -35,18 +36,26 @@ size_t work() {
...
                                    {"reopen_interval", "opi-903"}}),
                std::exception);
 
   // configure properly
-  logging::Configure(
-      {{"type", "file"}, {"file_name", "test/thread_file_log_test.log"}, {"reopen_interval", "1"}});
+  logging::Configure({{"type", "file"},
+                      {"file_name", base_path},
+                      {"reopen_interval", "1"},
+                      {"max_file_size", "4000"}, // Small size to trigger rolling
```
Comment (MohamedM216) [inline test/logging.cc:57]: I increased the size slightly to ensure the first file is large enough to hold the logs from all threads.

**Review by nilsnolde — APPROVED** [review]
